tokio/src/io/async_fd.rs

```diff
@@ -24,8 +24,11 @@ use std::{task::Context, task::Poll};
 ///
 /// The inner object is required to implement [`AsRawFd`]. This file descriptor
 /// must not change while [`AsyncFd`] owns the inner object. Changing the file
-/// descriptor results in unspecified behavior in the IO driver, which may
-/// include breaking notifications for other sockets/etc.
+/// descriptor through the underlying [`AsRawFd::as_raw_fd`] results
+/// in unspecified behavior in the IO driver, which may include breaking
+/// notifications for other sockets/etc. That is not to say that simply using
+/// a clone of the inner object would be problematic if that is a supported
+/// thing to do.
```

Comment:
I don't really like this wording. Some notes:

 * The fd is not changed _through_ `as_raw_fd`.
 * I'm not sure we have to expand on the kinds of ways it may break.
 * I don't want to make promises around clones of the fd. Those can easily get you in trouble too.

Reply:
Isn't the point that `as_raw_fs` exposes the actual raw file descriptor?

If clones of `File` might cause UB, then we have a problem (and probably `new` should be unsafe).

Notwithstanding the huge number of ways in which things can break, there is a definitely a need to provide more guidance around common use cases.

Reply:
To clarify the first point, if the raw file fd wasn't exposed, would there be a problem?

Reply:
I mean, the `AsyncFd` access the fd through `as_raw_fd`, but you don't change it using `as_raw_fd`. And the kinds of trouble are _unspecified_ behavior, not _undefined_ behavior. The second point was referring to "which may include breaking notifications for other sockets/etc."

Reply:
To clarify, the problem happens when `as_raw_fd` returns different values when called multiple times, but you don't _make it_ return different values _through_ `as_raw_fd` — it doesn't provide mutable access to the fd.

Reply:
It's a statement about the implementer of `AsRawFd` - if that does something strange, then you have a problem? I'm not sure of the right way of phrasing "done by but not done right here" which was the intention of "through the". How about "Changing the file descriptor through an unexpected implementation of `AsRawFd`..."?

Perhaps you could clarify the difference between unspecified behaviour and undefined behaviour...?

Reply:
One possible way to word it would be:

> The inner type must always return the same file descriptor, even if `as_raw_fd` is called multiple times. Violating this requirement can cause unspecified behavior.

Unspecified behavior means it does weird stuff. Undefined behavior means unsafe code gone wrong.

Reply:
Or maybe reverse the sentence.

> The `as_raw_fd` method on the inner type must always return the same file descriptor when called multiple times.

Reply:
I like the last one. I'll do that...